api/configuration/v1alpha1/kongsni_types.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2024 Kong, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	konnectv1alpha1 "github.com/kong/kubernetes-configuration/api/konnect/v1alpha1"
+)
+
+// KongSNI is the schema for SNI API which defines a Kong SNI.
+//
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:scope=Namespaced
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Programmed",description="The Resource is Programmed on Konnect",type=string,JSONPath=`.status.conditions[?(@.type=='Programmed')].status`
+// +kubebuilder:validation:XValidation:rule="size(self.spec.certificateRef.name) > 0", message="spec.certificateRef.name is required and cannot be empty"
+// +kubebuilder:validation:XValidation:rule="oldSelf.spec.certificateRef == self.spec.certificateRef", message="spec.certificateRef is immutable"
+type KongSNI struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec KongSNISpec `json:"spec"`
+
+	// +kubebuilder:default={conditions: {{type: "Programmed", status: "Unknown", reason:"Pending", message:"Waiting for controller", lastTransitionTime: "1970-01-01T00:00:00Z"}}}
+	Status KongSNIStatus `json:"status,omitempty"`
+}
+
+// KongSNIAPISpec defines specification of an SNI.
+type KongSNIAPISpec struct {
+	// Name is the name of the SNI. Required and must be a host or wildcard host.
+	// +kubebuilder:validation:MinLength=1
+	Name string `json:"name"`
+
+	// Tags is an optional set of strings associated with the SNI for grouping and filtering.
+	Tags []string `json:"tags,omitempty"`
+}
+
+// KongSNISpec defines specification of a Kong SNI.
+type KongSNISpec struct {
+	// CertificateRef is the reference to the certificate to which the KongSNI is attached.
+	CertificateRef TargetRef `json:"certificateRef"`
```

Comment:
`TargetRef` is a type defined for `KongPluginBinding`: https://github.com/Kong/kubernetes-configuration/blob/f963a53f87ac0b4cf2491f8de9d9ea9e5c464955/api/configuration/v1alpha1/kongpluginbinding_types.go#L127 We need to decide whether to make it reusable (and extract it to a separate file perhaps and rename to something more suitable the generic use case, e.g. `KongObjectRef`) or make such types separate (change the `TargetRef` to `KongPluginTargetRef` and add `KongSNICertificateRef`). I'd vote for the latter as it gives us the possibility to change their schema independently, but I'm also fine with reusing with a rename.

Reply:
I'm in favor of adding generic type for this like proposed. `KongObjectRef` sounds good to me.

Reply:
Extracted `KongObjectRef`. Would submit a separate PR to change other CRDs to use the type (and maybe also add `KongObjectRefWithGroupKind` ?)